### PR TITLE
Hamdi/[Docker+refactor]

### PR DESCRIPTION
**Pull request by hamdiallam**
Another way to deploy the plasma daemon using docker. Removes the need for the rebuild the same binary over again.

Refactors
1. plasmad and plasmacli under /cmd
2. Removed circular dependencies in the prove command
     a. Reuse `eth.InitConn` and `eth.InitPlasma` in `plasmacli`
3. Removed func init() in plasmacli. Each command adds all of it's subcommands in its `root.go` command.
3. HasTxExited in the eth module also returns an error if there was a failure
     - we still want to censor but want a better error returned to the client instead of a false positive

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Simple usage with a mounted data directory:
```
Comment (colin-axner) [inline Dockerfile]: Is this docker file also going to install a geth full node?

Comment (hamdiallam) [reply Dockerfile]: it will not. configuration file must have the endpoint. 

Could include a docker-compose that deploys a geth node

File: cmd/plasmacli/cmd/eth/challenge.go [hunk 2]
```diff
@@ -79,7 +81,7 @@ Usage:
 			var result *tm.ResultTx
 			result, confirmSignatures, err = getProof(owner, challengingPos)
 			if err != nil {
-				fmt.Errorf("failed to retrieve exit information: { %s }", err)
+				return fmt.Errorf("failed to retrieve exit information: { %s }", err)
```
Comment (colin-axner) [inline cmd/plasmacli/cmd/eth/challenge.go]: nice, this was fixed in store pr too

File: eth/plasma.go [hunk 3]
```diff
@@ -122,22 +140,24 @@ func (plasma *Plasma) CommitPlasmaHeaders(ctx sdk.Context, plasmaStore store.Pla
...
 
-// GetDeposit checks the existence of a deposit nonce
-func (plasma *Plasma) GetDeposit(plasmaBlock *big.Int, nonce *big.Int) (plasmaTypes.Deposit, *big.Int, bool) {
-	deposit, err := plasma.contract.Deposits(nil, nonce)
+// GetDeposit checks the existence of a deposit nonce. The state is synchronized with the provided `plasmaBlockHeight. The deposit
+// must have occured before or at the same pegged ethereum block as `plasmaBlockHeight`.
+func (plasma *Plasma) GetDeposit(plasmaBlockHeight *big.Int, nonce *big.Int) (plasmaTypes.Deposit, *big.Int, bool) {
+	deposit, err := plasma.Deposits(nil, nonce)
 	if err != nil {
 		// TODO: log the error
```
Comment (colin-axner) [inline eth/plasma.go]: todo should be taken out now?

File: eth/plasma.go [hunk 4]
```diff
@@ -187,30 +208,39 @@ func (plasma *Plasma) HasTxBeenExited(plasmaBlock *big.Int, position plasmaTypes
...
 
-	ethBlockNum, err := plasma.ethBlockPeg(plasmaBlock)
-	if err != nil {
-		plasma.logger.Error(fmt.Sprintf("could not get associated ETH Block for plasma block %d: %s", plasmaBlock.Int64(), err.Error()))
-		return true
+	// `Pending` or `Challenged` stateee
+	exited := e.State == 1 || e.State == 3
+
+	// synchronize with the correct ethereum state
+	if plasmaBlockHeight != nil {
```
Comment (colin-axner) [inline eth/plasma.go]: when is this nil?

**Comment by colin-axner** [comment]
please remove plasmad bin, and plasmacli bin if its there

File: eth/plasma.go (hunk 4 above)
Comment (hamdiallam) [reply eth/plasma.go]: if it's nil, disregards the ethereum block peg and checks against the latest state. This is used in plasmacli

**Comment by hamdiallam** [comment]
> please remove plasmad bin, and plasmacli bin if its there

oops, removing rn

File: README.md [hunk 5]
```diff
@@ -5,8 +5,6 @@
 [![Discord](https://img.shields.io/badge/discord-join%20chat-blue.svg)](https://discord.gg/YTB5A4P)
 [![license](https://img.shields.io/github/license/FourthState/plasma-mvp-rootchain.svg)](https://github.com/FourthState/plasma-mvp-sidechain/blob/master/LICENSE)
 
-Implementation of [Minimum Viable Plasma](https://ethresear.ch/t/minimal-viable-plasma/426) compatible with our [rootchain contract](https://github.com/FourthState/plasma-mvp-rootchain)
```
Comment (colin-axner) [inline README.md]: Why was this taken out again?

File: cmd/plasmacli/cmd/eth/challenge.go [hunk 6]
```diff
@@ -14,14 +14,16 @@ import (
 	"strconv"
 )
 
-func init() {
```
Comment (colin-axner) [inline cmd/plasmacli/cmd/eth/challenge.go]: are you certain this works? I ran into issues where binding flags on the same name with defaults could override a flags input because of the ordering in which they were binded

File: cmd/plasmacli/cmd/eth/exit.go [hunk 7]
```diff
@@ -74,12 +76,12 @@ Transaction Exit Usage:
 			From:     auth.From,
 			Signer:   auth.Signer,
 			GasLimit: gasLimit,
-			Value:    big.NewInt(minExitBond),
```
Comment (colin-axner) [inline cmd/plasmacli/cmd/eth/exit.go]: why was this changed? seems harder to update now

File: cmd/plasmacli/cmd/eth/challenge.go (hunk 6 above)
Comment (hamdiallam) [reply cmd/plasmacli/cmd/eth/challenge.go]: is this about the --trust-node flag? It'll definitely have to be changed here as using this command would override the persistent cli --trust-node flag

File: cmd/plasmacli/cmd/eth/exit.go (hunk 7 above)
Comment (hamdiallam) [reply cmd/plasmacli/cmd/eth/exit.go]: ah, minExitBond was defined in another file and only used in this location so i just inlined it. I can move it back

File: README.md (hunk 5 above)
Comment (hamdiallam) [reply README.md]: not sure, i didn't do this directly. maybe some merge issue. I'll put it back in

File: cmd/plasmacli/subcmd/eth/root.go [hunk 8]
```diff
@@ -25,6 +25,8 @@ const (
 	sigsF      = "signatures"
 	trustNodeF = "trust-node"
 	txBytesF   = "tx-bytes"
+
+	mixExitBond = 200000
```
Comment (colin-axner) [inline cmd/plasmacli/subcmd/eth/root.go]: min*

File: cmd/plasmacli/subcmd/eth/finalize.go [hunk 9]
```diff
@@ -2,18 +2,20 @@ package eth
...
 	"github.com/spf13/viper"
 	"strconv"
 )
 
-func init() {
-	ethCmd.AddCommand(finalizeCmd)
+func FinalizeCmd() *cobra.Command {
 	finalizeCmd.Flags().BoolP(depositsF, "D", false, "indicate that deposit exits should be finalized")
 	finalizeCmd.Flags().StringP(gasLimitF, "g", "240000", "gas limit for ethereum transaction")
+	viper.BindPFlags(finalizeCmd.Flags())
```
Comment (colin-axner) [inline cmd/plasmacli/subcmd/eth/finalize.go]: move to runE below

File: cmd/plasmacli/cmd/eth/challenge.go (hunk 6 above)
Comment (colin-axner) [reply cmd/plasmacli/cmd/eth/challenge.go]: Tested and my hunch was right:
- pulled this branch
- added print of `fmt.Printf(viper.GetString(gasLimitF))` to first line in challenge.go of runE
- ran plasmacli challenge cmd and `240000` was printed, either `300000` was expected or the `300` since I used `-g 300` on the cmd line.

Here is the fix, don't bind flags until the runE is being ran, otherwise binding by other flags will override the flag used on cmd line

File: cmd/plasmacli/subcmd/eth/query/balance.go [hunk 10]
```diff
@@ -23,7 +23,6 @@ Usage:
 	plasmacli eth query balance --address <address>`,
 	Args: cobra.MaximumNArgs(1),
 	RunE: func(cmd *cobra.Command, args []string) (err error) {
-		viper.BindPFlags(cmd.Flags())
```
Comment (colin-axner) [inline cmd/plasmacli/subcmd/eth/query/balance.go]: add back
